pkg/validation/usage_groups.go

```diff
@@ -0,0 +1,187 @@
+		[]string{"reason", "tenant", "usage_group"},
+	)
+)
+
+type UsageGroupConfig struct {
+	TenantID string
+	config   map[string][]*labels.Matcher
+}
+
+func (c *UsageGroupConfig) GetUsageGroups(lbls phlaremodel.Labels) UsageGroupMatch {
+	match := UsageGroupMatch{
+		tenantID: c.TenantID,
+	}
+
+	for name, matchers := range c.config {
+		if matchesAll(matchers, lbls) {
+			match.names = append(match.names, name)
+		}
+	}
+
+	return match
+}
+
+func (c *UsageGroupConfig) UnmarshalYAML(value *yaml.Node) error {
+	m := make(map[string]string)
+	err := value.DecodeWithOptions(&m, yaml.DecodeOptions{
+		KnownFields: true,
+	})
+	if err != nil {
+		return fmt.Errorf("malformed usage group config: %w", err)
+	}
+
+	*c, err = NewUsageGroupConfig(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *UsageGroupConfig) UnmarshalJSON(bytes []byte) error {
+	m := make(map[string]string)
+	err := json.Unmarshal(bytes, &m)
+	if err != nil {
+		return fmt.Errorf("malformed usage group config: %w", err)
+	}
+
+	*c, err = NewUsageGroupConfig(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type UsageGroupMatch struct {
+	tenantID string
+	names    []string
+}
+
+func (m UsageGroupMatch) CountReceivedBytes(profileType string, n int64) {
+	if len(m.names) == 0 {
+		usageGroupReceivedDecompressedBytes.WithLabelValues(profileType, m.tenantID, noMatchName).Add(float64(n))
+		return
+	}
+
+	for _, name := range m.names {
+		usageGroupReceivedDecompressedBytes.WithLabelValues(profileType, m.tenantID, name).Add(float64(n))
+	}
+}
+
+func (m UsageGroupMatch) CountDiscardedBytes(reason string, n int64) {
+	if len(m.names) == 0 {
+		usageGroupDiscardedBytes.WithLabelValues(reason, m.tenantID, noMatchName).Add(float64(n))
+		return
+	}
+
+	for _, name := range m.names {
+		usageGroupDiscardedBytes.WithLabelValues(reason, m.tenantID, name).Add(float64(n))
+	}
+}
+
+func NewUsageGroupConfig(m map[string]string) (UsageGroupConfig, error) {
+	if len(m) > maxUsageGroups {
+		return UsageGroupConfig{}, fmt.Errorf("maximum number of usage groups is %d, got %d", maxUsageGroups, len(m))
+	}
+
+	config := UsageGroupConfig{
+		config: make(map[string][]*labels.Matcher),
+	}
+
+	for name, matchersText := range m {
+		name = strings.TrimSpace(name)
+		if name == "" {
+			return UsageGroupConfig{}, fmt.Errorf("usage group name cannot be empty")
+		}
+
+		// TODO(bryanhuhta): We should probably validate the usage group name
+		// is a valid label value.
+
+		matchers, err := parser.ParseMetricSelector(matchersText)
+		if err != nil {
+			return UsageGroupConfig{}, fmt.Errorf("failed to parse matchers for usage group %q: %w", name, err)
+		}
+
+		config.config[name] = matchers
+	}
+
+	return config, nil
+}
+
+func (o *Overrides) DistributorUsageGroups(tenantID string) *UsageGroupConfig {
+	config := o.getOverridesForTenant(tenantID).DistributorUsageGroups
+
+	// It should never be nil, but check just in case!
+	if config == nil {
+		config = &UsageGroupConfig{}
+	}
```

Comment:
With the current code, the config will never be nil. However, I think we should leave this check here so we treat a nil config like a empty config, in case the config could be nil at this point.